From ece5bf72b24bb5e8861aab9d0153fc81ecea04a5 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Fri, 28 Nov 2014 09:58:13 +0200
Subject: [PATCH 322/456] v4l: xilinx: remapper: Handle video core clock

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

The video core clock is mandatory, manage it explicitly by requesting
and enabling it at probe time and disabling it a removal time.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
(cherry picked from commit d973b664dcadde7a13d2b4d384e6cd0304ba9546)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 .../bindings/media/xilinx/xlnx,v-remapper.txt      |    4 ++++
 drivers/media/platform/xilinx/xilinx-remapper.c    |   12 +++++++++++-
 2 files changed, 15 insertions(+), 1 deletions(-)

diff --git a/Documentation/devicetree/bindings/media/xilinx/xlnx,v-remapper.txt b/Documentation/devicetree/bindings/media/xilinx/xlnx,v-remapper.txt
index c8dff38..cda02cb 100644
--- a/Documentation/devicetree/bindings/media/xilinx/xlnx,v-remapper.txt
+++ b/Documentation/devicetree/bindings/media/xilinx/xlnx,v-remapper.txt
@@ -8,6 +8,8 @@ Required properties:
 
 - compatible: Must be "xlnx,v-remapper".
 
+- clocks: Reference to the video core clock.
+
 - xlnx,video-width: Video pixel component width, as defined in video.txt.
 
 - #xlnx,s-components: Number of components per pixel at the input port
@@ -31,6 +33,8 @@ Example: RBG to xRGB remapper
 	remapper_0: remapper {
 		compatible = "xlnx,v-remapper";
 
+		clocks = <&clkc 15>;
+
 		xlnx,video-width = <8>;
 
 		#xlnx,s-components = <3>;
diff --git a/drivers/media/platform/xilinx/xilinx-remapper.c b/drivers/media/platform/xilinx/xilinx-remapper.c
index 0acdb92..e22e95c 100644
--- a/drivers/media/platform/xilinx/xilinx-remapper.c
+++ b/drivers/media/platform/xilinx/xilinx-remapper.c
@@ -10,6 +10,7 @@
  * published by the Free Software Foundation.
  */
 
+#include <linux/clk.h>
 #include <linux/module.h>
 #include <linux/of.h>
 #include <linux/platform_device.h>
@@ -464,6 +465,12 @@ static int xremap_probe(struct platform_device *pdev)
 	if (ret < 0)
 		return ret;
 
+	xremap->xvip.clk = devm_clk_get(xremap->xvip.dev, NULL);
+	if (IS_ERR(xremap->xvip.clk))
+		return PTR_ERR(xremap->xvip.clk);
+
+	clk_prepare_enable(xremap->xvip.clk);
+
 	/* Initialize V4L2 subdevice and media entity */
 	subdev = &xremap->xvip.subdev;
 	v4l2_subdev_init(subdev, &xremap_ops);
@@ -480,7 +487,7 @@ static int xremap_probe(struct platform_device *pdev)
 	subdev->entity.ops = &xremap_media_ops;
 	ret = media_entity_init(&subdev->entity, 2, xremap->pads, 0);
 	if (ret < 0)
-		return ret;
+		goto error;
 
 	platform_set_drvdata(pdev, xremap);
 
@@ -496,6 +503,7 @@ static int xremap_probe(struct platform_device *pdev)
 
 error:
 	media_entity_cleanup(&subdev->entity);
+	clk_disable_unprepare(xremap->xvip.clk);
 	return ret;
 }
 
@@ -507,6 +515,8 @@ static int xremap_remove(struct platform_device *pdev)
 	v4l2_async_unregister_subdev(subdev);
 	media_entity_cleanup(&subdev->entity);
 
+	clk_disable_unprepare(xremap->xvip.clk);
+
 	return 0;
 }
 
-- 
1.7.5.4

